### PR TITLE
Add some tracing log for query optimizer

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Major changes in this PR:

1. Add some tracing log during query optimization, see more details in `Optimizer::optimize` and `ApplyRuleTask::execute`
2. Add a new session variable `enable_optimizer_trace_log` to control whether output these log;
3. Rewrite `toString` method of some critical operator so that they can output more detail information.

The following link is the tracing log generated by tpc-h q3, irrelevant logs have been removed：
[TPC-H Q3 Log](https://raw.githubusercontent.com/silverbullet233/starrocks/add_optimizer_trace_log_tmp/tracing_demo)

for https://github.com/StarRocks/starrocks/issues/4195

### CODE REVIEW

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 25 / 229 (10.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalProjectOperator.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/GroupExpression.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalDistributionOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/OrderSpec.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHiveScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalFilterOperator.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalEsScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJDBCScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHudiScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalLimitOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalEsScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMysqlScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalProjectOperator.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalLimitOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJDBCScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHudiScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHiveScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|5|15|33.33%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceInfo.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|12|24|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerContext.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Group.java [hunk 1]
```diff
@@ -250,4 +250,20 @@ public String toString() {
         return sb.toString();
     }
 
+    public String toPrettyString() {
+        return toPrettyString("", "");
+    }
+
+    public String toPrettyString(String headlineIndent, String detailIndent) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Group.java:257]: Why not directly change the `toString` method?

**Comment by kangkaisen** [comment]
Give an example for this PR. you could use TPCH-Q3 sql.

**Comment by silverbullet233** [comment]
> Give an example for this PR. you could use TPCH-Q3 sql.

added in description

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 25 / 225 (11.11%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalProjectOperator.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/GroupExpression.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalDistributionOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/OrderSpec.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHiveScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalFilterOperator.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalEsScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJDBCScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHudiScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalLimitOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalEsScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMysqlScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalProjectOperator.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalLimitOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJDBCScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHudiScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHiveScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|5|17|29.41%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceInfo.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|12|24|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerContext.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by kangkaisen** [comment]
> > Give an example for this PR. you could use TPCH-Q3 sql.
> 
> added in description

seems too many info. I think we should ensure the query tracing log is readable

**Comment by silverbullet233** [comment]
> > > Give an example for this PR. you could use TPCH-Q3 sql.
> > 
> > 
> > added in description
> 
> seems too many info. I think we should ensure the query tracing log is readable

Yes, the main problem now is that Groups are displayed in a hierarchical structure, and many Groups will be repeatedly output, resulting in a lot of content. A possible way is to output the Group separately, and only display the GroupID in the hierarchical structure, but it is very troublesome to view this way. Do you have any better suggestions?

**Comment by silverbullet233** [comment]
I have updated my changes and the output demo based on tpch-q3. please take a look again when you are available @kangkaisen

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 27 / 273 (09.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalProjectOperator.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalTopNOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/GroupExpression.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/analysis/CompoundPredicate.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalDistributionOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/SubqueryRelation.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/UnionRelation.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/Relation.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/QueryStatement.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/analysis/SelectListItem.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/OrderSpec.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHiveScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/SelectRelation.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalFilterOperator.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalEsScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJDBCScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalHudiScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalLimitOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalTopNOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/analysis/GroupByClause.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/LimitElement.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/JoinRelation.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalEsScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMysqlScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalProjectOperator.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalLimitOperator.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJDBCScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/analysis/SelectList.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHudiScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHiveScanOperator.java|0|7|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|5|17|29.41%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceInfo.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|3|7|42.86%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|11|21|52.38%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerContext.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Review by HangyuanLiu — COMMENTED** [review]
Do not scatter the code in each data structure definition class, please refer to the implementation form of Explain.java

File: fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java [hunk 2]
```diff
@@ -39,15 +39,29 @@
...
 
 public class StatementPlanner {
+    private static final Logger LOG = LogManager.getLogger(StatementPlanner.class);
+
     public ExecPlan plan(StatementBase stmt, ConnectContext session) throws AnalysisException {
+        if (session.getSessionVariable().isEnableOptimizerTraceLog()) {
+            if (stmt instanceof QueryStatement) {
+                LOG.info("[TRACE QUERY {}] After Parse\n{}\n", session.getQueryId(), stmt.toString());
+            }
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:57]: I think you can define a method `getOptimizeTraceLevel()` in session, and return `INFO` if enable optimizer trace log else return `DEBUG`/`TRACE`, then you can code:
```
LOG.log(getOptimizeTraceLevel(), 'xxxxxxxxxxxxxxxx');
```

or another ways, define a OptimizerTracer class, to support log interface and replace your `OptimizerTraceInfo` class

File: fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java [hunk 3]
```diff
@@ -39,15 +39,29 @@
...
 import java.util.Map;
 import java.util.stream.Collectors;
 
 public class StatementPlanner {
+    private static final Logger LOG = LogManager.getLogger(StatementPlanner.class);
+
     public ExecPlan plan(StatementBase stmt, ConnectContext session) throws AnalysisException {
+        if (session.getSessionVariable().isEnableOptimizerTraceLog()) {
+            if (stmt instanceof QueryStatement) {
+                LOG.info("[TRACE QUERY {}] After Parse\n{}\n", session.getQueryId(), stmt.toString());
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:55]: Last `\n` is redundant

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/ApplyRuleTask.java [hunk 4]
```diff
@@ -58,8 +64,25 @@ public void execute() {
...
+                sb.append("New Expressions:\n");
+                if (targetExpressions.isEmpty()) {
+                    sb.append("Empty\n");
+                } else {
+                    for (int i = 0; i < targetExpressions.size(); i++) {
+                        sb.append(i + ":\n" + targetExpressions.get(i).explain());
+                    }
+                }
+                LOG.info(sb.toString());
+                traceInfo.recordAppliedRule(rule.toString());
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/ApplyRuleTask.java:84]: Better to put these into a function

File: fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java (hunk 2 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:57]: > I think you can define a method `getOptimizeTraceLevel()` in session, and return `INFO` if enable optimizer trace log else return `DEBUG`/`TRACE`, then you can code:
> 
> ```
> LOG.log(getOptimizeTraceLevel(), 'xxxxxxxxxxxxxxxx');
> ```
> 
> or another way, define a OptimizerTracer class, to support log interface and replace your `OptimizerTraceInfo` class I perfer way 2

Yes, unifying the output interface will be clearer and reduce a log of judgment logic, I will fix it

Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:57]: > I think you can define a method `getOptimizeTraceLevel()` in session, and return `INFO` if enable optimizer trace log else return `DEBUG`/`TRACE`, then you can code:
> 
> ```
> LOG.log(getOptimizeTraceLevel(), 'xxxxxxxxxxxxxxxx');
> ```
> 
> or another way, define a OptimizerTracer class, to support log interface and replace your `OptimizerTraceInfo` class I perfer way 2

I think it is necessary to keep `OtpmizerTraceInfo` class to facilitate the collection of trace information, I will add a new class to deal with the output problem

**Comment by silverbullet233** [comment]
I have refactored the interface of aseembling trace info string and log output, see more defailts in `OptimizerTraceUtil`. PTAL @Seaven @HangyuanLiu @liuyehcf @kangkaisen

File: fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java [hunk 5]
```diff
@@ -32,22 +32,33 @@
...
 import com.starrocks.sql.plan.PlanFragmentBuilder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
 public class StatementPlanner {
+    private static final Logger LOG = LogManager.getLogger(StatementPlanner.class);
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java:51]: unsed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 41 / 327 (12.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/GroupExpression.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/analysis/CompoundPredicate.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpression.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/OrderSpec.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/SelectRelation.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalLimitOperator.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/SelectList.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|11|265|04.15%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceInfo.java|4|11|36.36%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerContext.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|5|5|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 40 / 326 (12.27%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/GroupExpression.java|0|9|00.00%|
|:large_blue_circle:|com/starrocks/analysis/CompoundPredicate.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpression.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/OrderSpec.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/SelectRelation.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalLimitOperator.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/analysis/SelectList.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/Projection.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|11|265|04.15%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceInfo.java|4|11|36.36%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|10|10|100.00%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/ApplyRuleTask.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerContext.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|4|4|100.00%|

**Review by Seaven — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**